dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,607 @@
 package api
 
 import (
-	"k8s.io/apimachinery/pkg/api/resource"
+	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
 )
 
 // DeviceClass is a vendor or admin-provided resource that contains
-// contraint and configuration information. Essentially, it is a re-usable
-// collection of predefined data that device claims may use.
+// device configuration and requirements. It can be referenced in
+// the device requests of a claim to apply these presets.
 // Cluster scoped.
 type DeviceClass struct {
-	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata,omitempty"`
-
-	Spec DeviceClassSpec `json:"spec,omitempty"`
-}
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
 
-// DeviceClassSpec provides either a selector to find other
-// classes, or the specific constraints defining this class.
-type DeviceClassSpec struct {
+	// Only nodes matching the selector will be considered by the scheduler
+	// when trying to find a Node that fits a Pod when that Pod uses
+	// a ResourceClaim that has not been allocated yet.
+	//
+	// Setting this field is optional. If unset, all nodes are candidates.
+	// It is ignored when the claim does not use a control plane controller
+	// for allocation.
+	//
+	// This is an alpha field and requires enabling the DRAControlPlaneController
+	// feature gate.
+	//
+	// +optional
+	SuitableNodes *v1.NodeSelector `json:"suitableNodes,omitempty" protobuf:"bytes,4,opt,name=suitableNodes"`
 
-	// One of selector and classCriteria must be populated.
+	// Config defines configuration parameters that apply to each request in a claim using this class.
+	// They are ignored while allocating the claim.
```

Comment:
> If you declare it as a listMap type with DriverName as the key, you get some amount of clarity and some support from the machinery about uniqueness.

The machinery does not automatically enforce uniqueness. That has to be hand-coded. The choice between `+listType=atomic` and `+listType=map` is more driven by "do we need to patch" and "who owns entries". I think `atomic` is more appropriate here.

I also wouldn't require "one config per driver". It might make sense for the same driver to have different config types that get provided separately.

Reply:
You can close this.  I commented below.